File: src/assets/src/hako_asset_impl.cpp
#include "hako_asset_impl.hpp"
#ifdef WIN32
#else
#include <unistd.h>
#ifdef ENABLE_HAKO_TIME_MEASURE
#include "hako_asset_impl_measure.hpp"
#endif /* ENABLE_HAKO_TIME_MEASURE */
#endif

HakoAssetType hako_asset_instance;

/***********************
 * simulation control
 ***********************/
static hako::asset::PduReader* create_reader(const nlohmann::json &reader_json)
{
    hako::asset::PduReader* reader = new hako::asset::PduReader{
        reader_json["type"],
        reader_json["org_name"],
        reader_json["name"],
        reader_json["channel_id"],
        reader_json["pdu_size"],
        reader_json["write_cycle"],
    };
    HAKO_ASSET_ASSERT(reader != nullptr);
    return reader;
}
static hako::asset::PduWriter* create_writer(const nlohmann::json &writer_json)
{
    hako::asset::PduWriter* writer = new hako::asset::PduWriter{
        writer_json["type"],
        writer_json["org_name"],
        writer_json["name"],
        writer_json["write_cycle"],
        writer_json["channel_id"],
        writer_json["pdu_size"],
        writer_json["method_type"]
    };
    HAKO_ASSET_ASSERT(writer != nullptr);
    return writer;
}
static void hako_asset_impl_parse_robots(bool is_plant)
{
    const json& robots_json = hako_asset_instance.param["robots"];
    
    for (const auto& robot_json : robots_json) {
        hako::asset::Robot* robot = new hako::asset::Robot;
        robot->name = robot_json["name"];
        if (robot_json.find("shm_pdu_writers") != robot_json.end()) {
            const json& pdu_readers_json = robot_json["shm_pdu_writers"];
            for (const auto& reader_json : pdu_readers_json) {
                if (is_plant) {
                    robot->pdu_writers.push_back(*create_writer(reader_json));
                }
                else {
                    robot->pdu_readers.push_back(*create_reader(reader_json));
                }
            }
        } else {
            // nothing to do
        }
        if (robot_json.find("rpc_pdu_writers") != robot_json.end()) {
            const json& pdu_readers_json = robot_json["rpc_pdu_writers"];
            for (const auto& reader_json : pdu_readers_json) {
                if (is_plant) {
                    robot->pdu_writers.push_back(*create_writer(reader_json));
                }
                else {
                    robot->pdu_readers.push_back(*create_reader(reader_json));
                }
            }
        } else {
            // nothing to do
        }
        if (robot_json.find("shm_pdu_readers") != robot_json.end()) {
            const json& pdu_writers_json = robot_json["shm_pdu_readers"];
            for (const auto& writer_json : pdu_writers_json) {
                if (is_plant) {
                    robot->pdu_readers.push_back(*create_reader(writer_json));
                }
                else {
                    robot->pdu_writers.push_back(*create_writer(writer_json));
                }
            }
        } else {
            // nothing to do
        }
        if (robot_json.find("rpc_pdu_readers") != robot_json.end()) {
            const json& pdu_writers_json = robot_json["rpc_pdu_readers"];
            for (const auto& writer_json : pdu_writers_json) {
                if (is_plant) {
                    robot->pdu_readers.push_back(*create_reader(writer_json));
                }
                else {
                    robot->pdu_writers.push_back(*create_writer(writer_json));
                }
            }
        } else {
            // nothing to do
        }        
        hako_asset_instance.robots.push_back(*robot);
    }
}

bool hako_asset_impl_init(const char* asset_name, const char* config_path, hako_time_t delta_usec, bool is_plant)
{
    hako_asset_instance.is_initialized = false;
    hako_asset_instance.external_use = false;
    std::ifstream ifs(config_path);
    
    if (!ifs.is_open()) {
        std::cerr << "Error: Failed to open config file." << std::endl;
        return false;
    }

    hako_asset_instance.asset_name_str = asset_name;
    hako_asset_instance.delta_usec = delta_usec;
    hako_asset_instance.current_usec = 0;
    try {
        hako_asset_instance.param = json::parse(ifs);
        hako_asset_impl_parse_robots(is_plant);
    } catch (const json::exception& e) {
        std::cerr << "JSON parsing error: " << e.what() << std::endl;
        return false;
    }
    hako_asset_instance.hako_asset = hako::create_asset_controller();
    if (hako_asset_instance.hako_asset == nullptr) {
        std::cerr << "ERROR: Not found hako-master on this PC" << std::endl;
        return false;
    }
    hako_asset_instance.hako_sim = hako::get_simevent_controller();
    if (hako_asset_instance.hako_sim == nullptr) {
        std::cerr << "ERROR: Not found hako-master on this PC" << std::endl;
        return false;
    }
    bool ret = hako_asset_instance.hako_asset->asset_register_polling(asset_name);
    if (ret == false) {
        std::cerr << "ERROR: Can not register asset" << std::endl;
        return false;
    }
    // PDUのチャネルを作成する
    for (const hako::asset::Robot& robot : hako_asset_instance.robots) {
        for (const hako::asset::PduWriter& writer : robot.pdu_writers) {
            std::cout << "Robot: " << robot.name << ", PduWriter: " << writer.name << std::endl;
            std::cout << "channel_id: " << writer.channel_id << " pdu_size: " << writer.pdu_size << std::endl;
            bool err = hako_asset_instance.hako_asset->create_pdu_lchannel(
                robot.name,
                writer.channel_id, 
                writer.pdu_size
            );
            if (err == false) {
                std::cerr << "ERROR: Can not create_pdu_channel()" << std::endl;
                return false;
            }            
        }
    }
#ifdef ENABLE_HAKO_TIME_MEASURE
    //create buffer
    hako_asset_instance.measure_vp = hako_asset_impl_measure_create_csv(asset_name);
    HAKO_IMPL_ASSERT(hako_asset_instance.measure_vp != nullptr);
#endif /* ENABLE_HAKO_TIME_MEASURE */
    hako_asset_instance.is_initialized = true;
    return true;
}
bool hako::asset::hako_asset_get_pdus(std::vector<hako::asset::Robot> &robots)
{
    if (!hako_asset_instance.is_initialized) {
        std::cerr << "ERROR: hako_asset_instance is not initialized." << std::endl;
        return false;
    }

    robots = hako_asset_instance.robots;
    return true;
}

bool hako_asset_impl_initialize_for_external()
{
    if (hako_asset_instance.is_initialized) {
        return false;
    }
    hako_asset_instance.is_initialized = false;
    hako_asset_instance.external_use = true;
    hako_asset_instance.asset_name_str = "None";
    hako_asset_instance.delta_usec = 0;
    hako_asset_instance.current_usec = 0;

    hako_asset_instance.hako_asset = hako::create_asset_controller();
    if (hako_asset_instance.hako_asset == nullptr) {
        std::cerr << "ERROR: Not found hako-master on this PC" << std::endl;
        return false;
    }
    hako_asset_instance.hako_sim = hako::get_simevent_controller();
    if (hako_asset_instance.hako_sim == nullptr) {
        std::cerr << "ERROR: Not found hako-master on this PC" << std::endl;
        return false;
    }

    hako_asset_instance.is_initialized = true;
    return true;
}
int hako_asset_impl_pdu_create(const char *robo_name, HakoPduChannelIdType lchannel, size_t pdu_size)
{
    if (!hako_asset_instance.is_initialized) {
        return false;
    }
    std::string robotName = robo_name;
    bool err = hako_asset_instance.hako_asset->create_pdu_lchannel(robotName, lchannel, pdu_size);
    if (err == false) {
        std::cerr << "ERROR: Can not create_pdu_channel()" << std::endl;
        return false;
    }    
    return true;
}
bool hako_asset_impl_register_callback(const hako_asset_callbacks_t* callback)
{
    if (hako_asset_instance.external_use) {
        return false;
    }
    hako_asset_instance.callback = callback;
    return true;
}
HakoSimulationStateType hako_asset_impl_state()
{
    return hako_asset_instance.hako_sim->state();
}
static bool hako_asset_impl_wait_state(HakoSimulationStateType target)
{
    do {
        HakoSimulationStateType curr = hako_asset_instance.hako_sim->state();
        if (curr == target) {
            break;
        }
        usleep(HAKO_ASSET_WAIT_TIME_USEC);
    } while (true);
    return true;
}

static bool hako_asset_impl_wait_event(HakoSimulationAssetEventType target)
{
    bool target_event_is_occureed = false;
    while (target_event_is_occureed == false) {
        auto event = hako_asset_instance.hako_asset->asset_get_event(hako_asset_instance.asset_name_str);
        target_event_is_occureed = (target == event);
        switch (event) {
            case HakoSimAssetEvent_None:
                usleep(HAKO_ASSET_WAIT_TIME_USEC);
                break;
            case HakoSimAssetEvent_Start:
                hako_asset_instance.hako_asset->start_feedback(hako_asset_instance.asset_name_str, true);
                break;
            case HakoSimAssetEvent_Stop:
                hako_asset_instance.hako_asset->stop_feedback(hako_asset_instance.asset_name_str, true);
                break;
            case HakoSimAssetEvent_Reset:
                if (hako_asset_instance.callback != NULL) {
                    if (hako_asset_instance.callback->on_reset != NULL) {
                        hako_asset_instance.callback->on_reset(nullptr);
                    }
                }
                hako_asset_instance.current_usec = 0;
                hako_asset_instance.hako_asset->reset_feedback(hako_asset_instance.asset_name_str, true);
                break;
            default:
                std::cerr << "ERROR: hako_asset_impl_wait_event() unkonwn event= " << event << std::endl;
                return false;
        }
    }
    return true;
}

static bool hako_asset_impl_wait_pdu_created()
{
    do {
        bool result = hako_asset_instance.hako_asset->is_pdu_created();
        if (result) {
            break;
        }
        usleep(HAKO_ASSET_WAIT_TIME_USEC);
    } while (true);
    return true;
}

bool hako_asset_impl_wait_running(void)
{
    std::cout << "WAIT START" << std::endl;
    if (hako_asset_impl_wait_event(HakoSimAssetEvent_Start) == false) {
        return false;
    }
    std::cout << "WAIT RUNNING" << std::endl;
    if (hako_asset_impl_wait_state(HakoSim_Running) == false) {
        return false;
    }
    std::cout << "PDU CREATED" << std::endl;
    if (hako_asset_impl_wait_pdu_created() == false) {
        return false;
    }
    if (hako_asset_instance.callback != NULL) {
        if (hako_asset_instance.callback->on_initialize != NULL) {
            hako_asset_instance.callback->on_initialize(nullptr);
        }
    }
    else {
	// nothing to do
    }
    return true;
}
static bool hako_asset_impl_execute(void)
{
    hako_asset_instance.hako_asset->notify_simtime(hako_asset_instance.asset_name_str, hako_asset_instance.current_usec);
    HakoSimulationStateType curr = hako_asset_instance.hako_sim->state();
    if (curr != HakoSim_Running) {
        std::cout << "NOT RUNNING: curr = " << curr << std::endl;
        return false;
    }
    bool result = hako_asset_instance.hako_asset->is_pdu_created();
    if (result == false) {
        std::cout << "PDU IS NOT CREATED" << std::endl;
        return false;
    }
    result = hako_asset_instance.hako_asset->is_pdu_sync_mode(hako_asset_instance.asset_name_str);
    if (result == true) {
        std::cout << "SYNC MODE: true" << std::endl;
	hako_asset_instance.hako_asset->notify_write_pdu_done(hako_asset_instance.asset_name_str);
        return false;
    }
    result = hako_asset_instance.hako_asset->is_simulation_mode();
    if (result == false) {
        //std::cout << "NOT SIMULATION MODE" << std::endl;
        return false;
    }
    hako_time_t world_time = hako_asset_instance.hako_asset->get_worldtime();
    hako_time_t next_time = hako_asset_instance.current_usec + hako_asset_instance.delta_usec;
    if (next_time > world_time) {
        return false;
    }
    if (hako_asset_instance.callback != NULL) {
        if (hako_asset_instance.callback->on_simulation_step != NULL) {
            hako_asset_instance.callback->on_simulation_step(nullptr);
        }
    }
    hako_asset_instance.current_usec = next_time;
    //std::cout << "# current_usec = " << hako_asset_instance.current_usec << std::endl;
#ifdef ENABLE_HAKO_TIME_MEASURE
    //write csv
    hako_asset_impl_measure_write_csv(hako_asset_instance.measure_vp, world_time, hako_asset_instance.current_usec);
#endif /* ENABLE_HAKO_TIME_MEASURE */
    return true;
}
static void hako_asset_impl_pdus_write_done(void)
{
    for (const hako::asset::Robot& robot : hako_asset_instance.robots) {
        for (const hako::asset::PduWriter& writer : robot.pdu_writers) {
            std::cout << "INFO: hako_asset_impl_pdus_write_done() Robot: " << robot.name << ", PduWriter: " << writer.name << std::endl;
            std::cout << "channel_id: " << writer.channel_id << " pdu_size: " << writer.pdu_size << std::endl;
            bool is_called = false;
            if (is_called == false) {
                char * buffer = (char*) malloc(writer.pdu_size);
                HAKO_ASSET_ASSERT(buffer != NULL);
                memset(buffer, 0, writer.pdu_size);
                auto ret = hako_asset_impl_pdu_write(robot.name.c_str(), writer.channel_id, buffer, writer.pdu_size);
                HAKO_ASSET_ASSERT(ret == true);
                free(buffer);
            }
        }
    }

}
static bool hako_asset_impl_proc(void)
{
    while (hako_asset_impl_execute() == false) {
        HakoSimulationStateType curr = hako_asset_instance.hako_sim->state();
        if (curr != HakoSim_Running) {
#ifdef ENABLE_HAKO_TIME_MEASURE
            //close file
            hako_asset_impl_measure_flush_csv(hako_asset_instance.measure_vp);
#endif /* ENABLE_HAKO_TIME_MEASURE */
            std::cout << "WAIT STOP" << std::endl;
            auto ret = hako_asset_impl_wait_event(HakoSimAssetEvent_Stop);
            HAKO_ASSET_ASSERT(ret == true);
            std::cout << "WAIT RESET" << std::endl;
            ret = hako_asset_impl_wait_event(HakoSimAssetEvent_Reset);
            HAKO_ASSET_ASSERT(ret == true);
            // WAINT FOR RUNNING STATE
            //ret = hako_asset_impl_wait_running();
            //HAKO_ASSET_ASSERT(ret == true);
            return false;
        }
        else if (hako_asset_instance.hako_asset->is_pdu_sync_mode(hako_asset_instance.asset_name_str) == true) {
            hako_asset_impl_pdus_write_done();
        }
    }
    return true;
}

bool hako_asset_impl_step(hako_time_t increment_step)
{
    hako_time_t target_time_usec = hako_asset_instance.current_usec + (increment_step * hako_asset_instance.delta_usec);

    // RUNNING PROC
    //std::cout << "current_usec = " << hako_asset_instance.current_usec << std::endl;
    //std::cout << "target_time_usec = " << target_time_usec << std::endl;
    while (hako_asset_instance.current_usec < target_time_usec) {
        if (hako_asset_impl_proc() == false) {
            return false;
        }
    }
    return true;
}
/***********************
 * pdu io
 ***********************/

bool hako_asset_impl_pdu_read(const char* robo_name, HakoPduChannelIdType lchannel, char* buffer, size_t buffer_len)
{
    if (hako_asset_instance.external_use) {
        return hako_asset_instance.hako_asset->read_pdu_for_external(robo_name, lchannel, buffer, buffer_len);
    }
    bool ret = hako_asset_instance.hako_asset->read_pdu(hako_asset_instance.asset_name_str, robo_name, lchannel, buffer, buffer_len);
    if (ret == false) {
        return false;
    }
    return true;
}

bool hako_asset_impl_pdu_write(const char* robo_name, HakoPduChannelIdType lchannel, const char* buffer, size_t buffer_len)
{
    if (hako_asset_instance.external_use) {
        return hako_asset_instance.hako_asset->write_pdu_for_external(robo_name, lchannel, buffer, buffer_len);
    }
    bool ret = hako_asset_instance.hako_asset->write_pdu(hako_asset_instance.asset_name_str, robo_name, lchannel, buffer, buffer_len);
    if (ret == false) {
        return false;
    }
    hako_asset_instance.hako_asset->notify_write_pdu_done(hako_asset_instance.asset_name_str);
    return true;
}

hako_time_t hako_asset_impl_get_world_time()
{
    return hako_asset_instance.hako_asset->get_worldtime();
}

File: bindings/elixir_project/lib/impl/hako_asset_impl.ex
defmodule HakoAssetImpl do
  use Agent

  alias HakoAssetImpl
  alias HakoApi
  alias HakoSimevent
  import Jason
  @hako_asset_wait_time_usec 10000  # 1000 * 10 microseconds

  # HakoAssetインスタンスの初期化
  def start_link(_) do
    initial_state = %{
      external_use: false,
      is_initialized: false,
      asset_instance: %{
        asset_name: nil,
        config_path: nil,
        delta_usec: 0,
        current_usec: 0,
        robots: [],
        callback: nil,
        param: %{}
      }
    }
    Agent.start_link(fn -> initial_state end, name: __MODULE__)
  end

  # external_useフラグの設定と取得
  def set_external_use(state) do
    Agent.update(__MODULE__, fn instance -> %{instance | external_use: state} end)
  end

  def get_external_use() do
    Agent.get(__MODULE__, & &1.external_use)
  end

  # asset_instanceの設定と取得
  def set_asset_instance(new_asset_instance) do
    Agent.update(__MODULE__, fn instance -> %{instance | asset_instance: new_asset_instance} end)
  end

  def get_asset_instance() do
    Agent.get(__MODULE__, & &1.asset_instance)
  end

  # asset_instanceの初期化
  def reset_asset_instance() do
    Agent.update(__MODULE__, fn instance ->
      %{instance |
        asset_instance: %{
          asset_name: nil,
          config_path: nil,
          delta_usec: 0,
          current_usec: 0,
          callback: nil,
          robots: [],
          param: %{}
        },
        is_initialized: false
      }
    end)
  end

  # ロボットのPDUリーダーを作成
  defp create_reader(reader_json) do
    %{
      type: reader_json["type"],
      org_name: reader_json["org_name"],
      name: reader_json["name"],
      channel_id: reader_json["channel_id"],
      pdu_size: reader_json["pdu_size"],
      write_cycle: reader_json["write_cycle"]
    }
  end

  # ロボットのPDUライターを作成
  defp create_writer(writer_json) do
    %{
      type: writer_json["type"],
      org_name: writer_json["org_name"],
      name: writer_json["name"],
      write_cycle: writer_json["write_cycle"],
      channel_id: writer_json["channel_id"],
      pdu_size: writer_json["pdu_size"],
      method_type: writer_json["method_type"]
    }
  end

  # ロボットの設定を解析してPDUリーダー・ライターをセットアップ
  defp parse_robots(is_plant, robots_json) do
    Enum.map(robots_json, fn robot_json ->
      %{
        name: robot_json["name"],
        pdu_writers: parse_pdu_writers(is_plant, robot_json),
        pdu_readers: parse_pdu_readers(is_plant, robot_json)
      }
    end)
  end

  defp parse_pdu_writers(is_plant, robot_json) do
    pdu_writers = []

    if Map.has_key?(robot_json, "shm_pdu_writers") do
      pdu_writers = pdu_writers ++ parse_writer_or_reader(is_plant, robot_json["shm_pdu_writers"])
    end

    if Map.has_key?(robot_json, "rpc_pdu_writers") do
      pdu_writers = pdu_writers ++ parse_writer_or_reader(is_plant, robot_json["rpc_pdu_writers"])
    end

    pdu_writers
  end

  defp parse_pdu_readers(is_plant, robot_json) do
    pdu_readers = []

    if Map.has_key?(robot_json, "shm_pdu_readers") do
      pdu_readers = pdu_readers ++ parse_writer_or_reader(is_plant, robot_json["shm_pdu_readers"], false)
    end

    if Map.has_key?(robot_json, "rpc_pdu_readers") do
      pdu_readers = pdu_readers ++ parse_writer_or_reader(is_plant, robot_json["rpc_pdu_readers"], false)
    end

    pdu_readers
  end

  defp parse_writer_or_reader(is_plant, json_array, is_writer \\ true) do
    Enum.map(json_array, fn entry_json ->
      if is_plant do
        if is_writer do
          create_writer(entry_json)
        else
          create_reader(entry_json)
        end
      else
        if is_writer do
          create_reader(entry_json)
        else
          create_writer(entry_json)
        end
      end
    end)
  end
  defp create_pdu_channels(robots) do
    Enum.each(robots, fn robot ->
      Enum.each(robot.pdu_writers, fn writer ->
        IO.puts("Robot: #{robot.name}, PduWriter: #{writer.name}")
        IO.puts("channel_id: #{writer.channel_id}, pdu_size: #{writer.pdu_size}")

        case HakoApi.create_pdu_lchannel(robot.name, writer.channel_id, writer.pdu_size) do
          true ->
            :ok
          false ->
            IO.puts("ERROR: Can not create_pdu_channel()")
            {:error, "Failed to create PDU channel"}
        end
      end)
    end)
  end

  # Assetの初期化
  def init(asset_name, config_path, delta_usec, is_plant) do
    reset_asset_instance()

    asset_instance = get_asset_instance()

    if File.exists?(config_path) do
      {:ok, file_content} = File.read(config_path)
      parsed_json = Jason.decode!(file_content)

      robots = parse_robots(is_plant, parsed_json["robots"])

      updated_instance = %{
        asset_instance |
        asset_name: asset_name,
        config_path: config_path,
        delta_usec: delta_usec,
        current_usec: 0,
        param: parsed_json,
        robots: robots,
      }

      if HakoApi.init_asset() == false or HakoApi.simevent_init() == false do
        {:error, "Hako master or Simevent controller not found"}
      else
        set_asset_instance(updated_instance)

        # Assetの登録処理
        case HakoApi.register_asset(asset_name) do
          true ->
            # PDUチャネルの作成
            create_pdu_channels(updated_instance.robots)
            :ok
          false ->
            {:error, "Failed to register asset"}
        end
      end
    else
      {:error, "Config file not found"}
    end
  end

  # Callbackの登録
  def register_callback(callbacks) do
    asset_instance = get_asset_instance()

    if get_external_use() do
      {:error, "External use is enabled; cannot register callbacks"}
    else
      updated_instance = %{asset_instance | callback: callbacks}
      set_asset_instance(updated_instance)
      :ok
    end
  end

  # 外部使用用の初期化
  def initialize_for_external() do
    asset_instance = get_asset_instance()

    if asset_instance.is_initialized do
      {:error, "Asset is already initialized"}
    else
      updated_instance = %{
        asset_instance |
        is_initialized: true,
        external_use: true,
        asset_name: "None",
        delta_usec: 0,
        current_usec: 0,
      }

      set_asset_instance(updated_instance)
      :ok
    end
  end

  # Assetの状態を取得
  def get_state() do
    HakoSimevent.get_state()
  end

  # 指定された状態まで待機
  defp wait_state(target_state) do
    Stream.repeatedly(fn -> :ok end)
    |> Enum.each_while(fn _ ->
      curr_state = HakoSimevent.get_state()
      if curr_state == target_state do
        {:halt, :ok}
      else
        Process.sleep(@hako_asset_wait_time_usec / 1000)
        {:cont, :ok}
      end
    end)
  end
  # 指定されたイベントまで待機
  defp wait_event(target_event) do
    Stream.repeatedly(fn -> :ok end)
    |> Enum.each_while(fn _ ->
      asset_name = get_asset_instance().asset_name
      event = HakoApi.get_event(asset_name)

      cond do
        event == target_event -> {:halt, :ok}
        event == :start ->
          HakoApi.start_feedback(asset_name, true)
          {:cont, :ok}
        event == :stop ->
          HakoApi.stop_feedback(asset_name, true)
          {:cont, :ok}
        event == :reset ->
          callback = get_asset_instance().callback
          if callback && callback.on_reset do
            callback.on_reset(nil)
          end
          HakoApi.reset_feedback(asset_name, true)
          {:cont, :ok}
        true ->
          Process.sleep(@hako_asset_wait_time_usec / 1000)
          {:cont, :ok}
      end
    end)
  end

  # PDU作成まで待機
  defp wait_pdu_created() do
    Stream.repeatedly(fn -> :ok end)
    |> Enum.each_while(fn _ ->
      if HakoApi.is_pdu_created() do
        {:halt, :ok}
      else
        Process.sleep(@hako_asset_wait_time_usec / 1000)
        {:cont, :ok}
      end
    end)
  end
  defp wait_running() do
    IO.puts("WAIT START")

    # イベントが Start になるのを待機
    case wait_event(:start) do
      :ok -> :ok
      _ -> false
    end

    IO.puts("WAIT RUNNING")

    # 状態が Running になるのを待機
    case wait_state(:running) do
      :ok -> :ok
      _ -> false
    end

    IO.puts("PDU CREATED")

    # PDU が作成されるのを待機
    case wait_pdu_created() do
      :ok -> :ok
      _ -> false
    end

    # コールバック on_initialize を呼び出す
    callback = get_asset_instance().callback

    if callback && callback.on_initialize do
      callback.on_initialize(nil)
    end

    true
  end

  defp execute_simulation() do
    asset_instance = get_asset_instance()
    HakoApi.notify_simtime(asset_instance.asset_name, asset_instance.current_usec)

    cond do
      HakoSimevent.get_state() != :running ->
        false

      not HakoApi.is_pdu_created() ->
        false

      HakoApi.is_pdu_sync_mode(asset_instance.asset_name) ->
        HakoApi.notify_write_pdu_done(asset_instance.asset_name)
        false

      not HakoApi.is_simulation_mode() ->
        false

      asset_instance.current_usec >= HakoApi.get_worldtime() ->
        false

      asset_instance.callback && asset_instance.callback.on_simulation_step ->
        asset_instance.callback.on_simulation_step(nil)
        true

      true ->
        false
    end
  end
  defp pdus_write_done() do
    asset_instance = get_asset_instance()

    Enum.each(asset_instance.robots, fn robot ->
      Enum.each(robot.pdu_writers, fn writer ->
        IO.puts("INFO: hako_asset_impl_pdus_write_done() Robot: #{robot.name}, PduWriter: #{writer.name}")
        IO.puts("channel_id: #{writer.channel_id} pdu_size: #{writer.pdu_size}")

        # Bufferをゼロで埋める
        buffer = :binary.copy(<<0>>, writer.pdu_size)

        # PDU書き込みの実行
        case HakoApi.pdu_write(robot.name, writer.channel_id, buffer) do
          :ok -> :ok
          _ -> raise "PDU write failed"
        end
      end)
    end)
  end

  defp proc() do
    loop()
  end

  defp loop() do
    if execute_simulation() == false do
      curr = HakoSimevent.get_state()

      if curr != :running do
        IO.puts("WAIT STOP")
        case wait_event(:stop) do
          :ok -> :ok
          _ -> raise "Waiting for STOP event failed"
        end

        IO.puts("WAIT RESET")
        case wait_event(:reset) do
          :ok -> :ok
          _ -> raise "Waiting for RESET event failed"
        end

        false
      else
        if HakoApi.is_pdu_sync_mode(get_asset_instance().asset_name) do
          pdus_write_done()
        end
        # ループを再帰的に呼び出す
        loop()
      end
    else
      true
    end
  end
  def step(increment_step) do
    asset_instance = get_asset_instance()
    target_time_usec = asset_instance.current_usec + (increment_step * asset_instance.delta_usec)

    run_until_target_time(asset_instance.current_usec, target_time_usec)
  end

  defp run_until_target_time(current_usec, target_time_usec) when current_usec < target_time_usec do
    if proc() do
      asset_instance = get_asset_instance()
      run_until_target_time(asset_instance.current_usec, target_time_usec)
    else
      false
    end
  end

  defp run_until_target_time(_, _) do
    true
  end


end
